pyproject.toml

```diff
@@ -2,7 +2,7 @@
 requires = ["setuptools",
             "setuptools_scm>=6.2",
             "cython==0.29.34",
-            "oldest-supported-numpy",
+            "numpy==1.25",
```

Comment:
Maybe this should be >=1.25

Reply:
I wonder if this should perhaps also include <2 though I haven't had time to check if the backward compatibility is 'forever' or could break in 2.0

Reply:
"A major release is highly unusual (NumPy is still at version 1.xx) and if it happens it will likely indicate an ABI break"

I think we should perhaps do >=1.25,<2

Reply:
I thought someone once said that we should always do `==` pins on the build requirements? Isn't that why we do `==` for cython? Then why cython but not numpy?

Reply:
Agreed about the `>=1.25,<2.0` pin for Numpy, 
For Cython, Numpy uses `Cython>=0.29.34,<3.0`  which seems better to me than pinning a specific version.

Reply:
But how can we know numpy 1.26 won't drop things what we need?

Reply:
Yes, I think we can trust Numpy on this and go with `>=1.25,<2.0`.

Reply:
OK I updated the pins but I am not so sure about touching `get_extensions`, so I'll leave that to you.

Reply:
NumPy 2.0 breaks forward compatibility but provides backwards compatibility: A wheel compiled with 1.X cannot run with 2.X.  So you will have to compile wheels against 2.0 when it comes out.

I may introduce a `numpy2_compat` as a runtime dependency to run against 1.X, but I haven't made up my mind yet: it might be too much hassle for little payoff (also considering dev cycles).

I will note that the only real reason to add:
```
#define NPY_TARGET_VERSION NPY_1_21_API_VERSION
```
at this point is if you would actually want to use newer C-API, which you don't.

Reply:
> A wheel compiled with 1.X cannot run with 2.X.

That is bad news for downstream packages that pull dev wheels for both `numpy` and `astropy`. How do we generate a dev wheel of astropy that is compatible with dev wheel of numpy before numpy 2.0 comes out?

Reply:
Hmmmpf, I guess you will have to compile it against our dev wheel, you don't even have to pass tests on it to do that.

The problem is, only if you use the 2.x wheel, can I provide updated headers to ensure compatibility (and force you to remove API that is I want to stop dragging around).

Reply:
Sorry, somehow slipped me:  Right now, NumPy doesn't actually enforce this!  For precisely that reason.  As long as we just remove things for the most part, I am happy to gamble that the wheels are effectively ABI compatible (even if that is a lie).

There might come a day where we want to do something that makes this impossible, but I am not sure that will come...
As soon as there is an actual NumPy `rc1`, that would be enforced though.

Reply:
> you will have to compile it against our dev wheel

I guess it is theoretically possible for our unreleased 6.0.dev. But we cannot do that for release branches; so does this mean we have to pin numpy maxversion for v5.3.x ? Though I also have no seen any failures related to this in our CI set up. My head hurts. 😆

Reply:
Sorry, yes, just repeating in case it got lost:  The NumPy dev wheels do *not* actually enforce this, they just pretend they are still 1.x ABI (it's a lie, but who cares).  Maybe that will just stay that way, maybe I will find a big enough reason to break that "promise".